### DISCUSSION TITLE
Commercial Ecosystem

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#732] [General]
One of the aspects which made DNN very successful and provided longevity to the platform is the commercial ecosystem around it. Part of that commercial ecosystem was predicated on the ability of the platform to enable independent modules and themes which could be packaged and distributed by third party ISVs and installed by non-technical consumers. The other part of the commercial ecosystem was the availability of a marketplace to facilitate transactions between ISVs and consumers. The marketplace for DNN was a traditional e-commerce website where ISV's could list their products and consumers could purchase them. The limitation of this approach is that consumers had to physically visit the marketplace site, purchase a product, download the package to their local machine, upload the package to their site to install it, and then deal with any licensing concerns ( which was totally inconsistent as each ISV had to develop their own licensing solution ). A more modern approach would streamline this process so that the interaction is much more fluid and simplifies the process for the consumer - much like the "app store" experience they are familiar with on mobile devices. In order for Oqtane to be truly successful I believe it needs a fully integrated marketplace and standardized licensing solution that are part of the core framework. The current integration with Nuget is the first step in this direction in terms of seamless installation of packages. I would like to hear people's thoughts on this topic...

**Comment by markdav-is** [comment]
I think it's a great idea to have discoverable themes and modules that are installable thru the app.  I like the app-store metaphor as it's each on the consumer.  Those modules and themes will all need to be reviewed and vetted.  I've been breaking my site with bad modules.  You could take a % of the cost like the app store does, and that would go to fund the vetting process and maintaining the store.  I use Open Collective for transparent accounting, and I could help with establishing a sustainability model.  Of course, folks could in most cases work around the store if the modules are open-source, but they might like to support the project as well.

**Comment by PizzaConsole** [comment]
I think its a great idea and its already the direction in which I saw Oqtane going. I would love to just be a full-time Oqtane Module developer

**Comment by PizzaConsole** [comment]
Also, essentially what you are proposing is just wordpress rebuilt in blazor, which is totally fine.

**Reply by abukres** [reply]
In WordPress you still have to get a commercial plugin, a zip file, from the vendor's site and install it manually. I don't see how this is different than how DNN works. At least most commercial DNN modules are available from one market place. WordPress has many marketplaces.

**Comment by mikecasas** [comment]
What was the most popular/successful licensing model for DNN? Meaning you purchase Module A for $250, and then can use it on all your installations? Only on one installation? What about when I upgrade my module to V2, do I re-charge $250 or just $50 for an upgrade?

I know the quick answer will be that each ISV can develop their own plan, but just wondering what worked the best. Also what may work as the "store" initially releases as compared to a more mature ecosystem is probably different too.

Also, how will a module be protected from unauthorized use? When you install it, will there be a key that is activated, so others can't use the module too?

Very interesting discussion...

**Reply by mikecasas** [reply]
Maybe Oqtane could first be in the Azure Market Place, under the web frameworks? I know there is the Deploy to Azure button in the ReadMe, but this would provide more visibility.

https://azuremarketplace.microsoft.com/en-us/marketplace/apps/category/web?page=1&subcategories=web-app-frameworks

Instead of reinventing the wheel maybe try and integrate with something like the Azure Marketplace. I am not saying Azure is better than AWS (or any other cloud vendor), I am just more familiar with Azure.

**Comment by ADefWebserver** [comment]
Perhaps if the "Marketplace Framework" had three options:
* Free
* One-Time Payment
* Subscription

**Comment by larsontim12** [comment]
If it is truly an open source project then it needs to be built that to keep it going that way. Either you have to have the ability to have multiple stores for the users to choose for or a module can be installed that acts as a store which is customized for an enterprise. 

Since this is an application it needs to keep its focus on that and not lose focus of who is going to be using the application framework.

Just my thoughts no offense attended.

Tim

**Comment by abukres** [comment]
While the idea is nice, the infrastructure to support all kinds of licensing is big. That's why you had each module or theme vendor implement their own. You can have a license for a free version. A license per domain. A license per installation. Maybe a free license for localhost. A license for x number of domains. A license that expires after x days. A way to transfer a license to another domain. If the license server goes down, then it affects everyone. You want to deal with all this hassle? Are you thinking of having a standard UI inside Oqtane for licensing which vendors conform to a standard licensing API but each vendor will actually control the actual licensing infrastructure?

**Comment by kevind2000** [comment]
Currently Oqtane is a web application framework, not web content management application. As a modular web application framework based on Blazor, it ticks many boxes of excellent features. As full-stack C# developers, we are excited with the promise of Blazor as client-side application technology and see the need of modular development and dynamic composition of UI which Oqtane brings to us, along with other great features, like multi-tenants.

Based on this application framework, we expect to build all kinds of applications and the first one might be CMS, namely, Blazorized and Modernized DNN. Like DNN, the commercial potential is already there, now there might be different ways to explore it.  As a few already pointed it out, this requires extra and separate effort to bring it out as a "product". Personally, I see this would be a nature development of the Oqtane - take Orchard Core for an example, it has modular web application framework and CMS on top of the framework using the ASP.NET core technologies. 

Back to Oqtane as application framework, I hope it's continued to be developed to support .NET 5 and hopefully matured to .NET 6, i.e. 2021 LTS. By then, I wish Oqtane supports not only web client applications, but also mobile and desktop client applications.

**Comment by ghost** [comment]
Suppose I created a commercial module tomorrow and want to start selling it, what are my options now? I think I need to put up a website and sell it there. There's no marketplace now. 

Are developers going to wait for a marketplace to happen before they think of creating a commercial module? What if someone decided to create their own marketplace like how Snowcovered.com got traction for DNN.

**Reply by sbwalker** [reply]
It someone wanted to create/sell a commercial module right now they would need to come up with their own solution for both the e-commerce and licensing aspects. There is nothing stopping anyone from doing this - and this is exactly the same approach which was required before Snowcovered appeared for DNN. A fully integrated marketplace as I described in my original post needs to be a high priority for Oqtane 2.x.

**Reply by mikecasas** [reply]
@sbwalker do you plan on taking this on, or do you want someone else to spearhead it?

**Reply by sbwalker** [reply]
Yes I planned to take this is on... it was one of the motivations behind posting this thread. I have a design and a partial solution already.... focused initially on the licensing aspect.

**Reply by mikecasas** [reply]
Ok great. If you need help, the community is here.

**Reply by ghost** [reply]
I assume your license solution is an option among licensing options developers would want to choose from. I think some developers will want to use their own licensing solution for whatever reason.

**Reply by sbwalker** [reply]
@thenrich2009 As many people already identified in this thread there are too many variations in business models to consider a one size fits all solution.... so developers will need to have the freedom to roll their own licensing solution. That being said, in the DNN ecosystem there were a common set of requirements which could handle 80% of the use cases, so this is what the default solution will target. The main goal is to provide a streamlined and intuitive user experience.

**Comment by ADefWebserver** [comment]
For now a commercial developer can package their custom **Oqtane** module(s) in a custom **Oqtane** distribution using the method described here: [Creating a Custom Distribution of Blazor Oqtane Using Site Templates](https://blazorhelpwebsite.com/ViewBlogPost/41). 

They can even create their own "**Deploy To Azure**" button:

![image_91bcbb88-e115-4703-816c-7878c97e22a0](https://user-images.githubusercontent.com/1857799/97790416-bc60e880-1b85-11eb-8976-1c01c24181ed.png)

**Reply by sbwalker** [reply]
This is a good point @ADefWebserver. You can definitely create custom distributions already :)
